### PR TITLE
Include cause that was used to notify the promise when logging an fai…

### PR DESCRIPTION
**Pull request by normanmaurer**
…led try to notify it.

Motivation:

When a promise is notified that was already added to the ChannelOutboundBuffer and we try to notify it later on we only see a warning that it was notified before. This is often not very useful as we have no idea where it was notified at all. We can do better in case it was failed before (which is most of the times the case) and just also log the cause that was used for it.

Modifications:

Add the cause that was used to notify the promise when we fail to notify it as part of the ChannelOutboundBuffer.

Result:

Easier to debug user errors.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler @trustin PTAL... I found this very useful when try to understand why a promise was already done (which in fact was my fault at the end ;))

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 1]
```diff
@@ -671,13 +674,41 @@ public void run() {
...
+                logger.warn("Failed to mark a promise as failure because it's done already: {} => {} ",
+                        promise, stackTraceToString(err), cause);
+            }
+        }
+    }
+
+    private static String stackTraceToString(Throwable cause) {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        PrintStream pout = new PrintStream(out);
+        cause.printStackTrace(pout);
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: CRITICAL Use a logger to log this exception. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1148)

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: nope ;)

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 2]
```diff
@@ -671,13 +674,41 @@ public void run() {
...
+            }
         }
     }
 
     private static void safeFail(ChannelPromise promise, Throwable cause) {
         if (!(promise instanceof VoidChannelPromise) && !promise.tryFailure(cause)) {
-            logger.warn("Failed to mark a promise as failure because it's done already: {}", promise, cause);
+            Throwable err = promise.cause();
+            if (err == null) {
+                logger.warn("Failed to mark a promise as failure because it's done already: {} ", promise, cause);
```
Comment (nmittler) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: Do you have the correct number of parameters here and below?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: yes as the "cause" is the actual exception that is passed to the logger.

**Comment by nmittler** [comment]
@normanmaurer Just one comment ... otherwise LGTM

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 3]
```diff
@@ -671,13 +674,41 @@ public void run() {
...
+            } else {
+                logger.warn("Failed to mark a promise as failure because it's done already: {} => {} ",
+                        promise, stackTraceToString(err), cause);
+            }
+        }
+    }
+
+    private static String stackTraceToString(Throwable cause) {
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: should this go into some internal utility class somewhere?

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 4]
```diff
@@ -671,13 +674,41 @@ public void run() {
...
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        PrintStream pout = new PrintStream(out);
+        cause.printStackTrace(pout);
+        pout.flush();
+        try {
+            return new String(out.toByteArray());
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: is taking the default system encoding what we want here?

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 2 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: nit: remove space after `{}` here and below?

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 4 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: I would say so...

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 3 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: as we not need it anywhere else I would just leave it here.

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 2 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: +1

**Comment by Scottmitch** [comment]
few questions and lgtm

**Comment by trustin** [comment]
Could we make the log message a little bit clearer so that which one was set and which one failed to be set?

**Comment by normanmaurer** [comment]
Sure... Let me do it and ping you once done

**Comment by normanmaurer** [comment]
@trustin PTAL if it's better now.

**Comment by normanmaurer** [comment]
@trustin ping...

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 5]
```diff
@@ -671,13 +674,41 @@ public void run() {
 
     private static void safeSuccess(ChannelPromise promise) {
         if (!(promise instanceof VoidChannelPromise) && !promise.trySuccess()) {
-            logger.warn("Failed to mark a promise as success because it is done already: {}", promise);
+            Throwable err = promise.cause();
+            if (err == null) {
+                logger.warn("Failed to mark a promise as success because it is succeed already: {}", promise);
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: - is succeed -> has succeeded

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 6]
```diff
@@ -671,13 +674,41 @@ public void run() {
...
+            } else {
+                logger.warn("Failed to mark a promise as success because it is failed already: {} with cause {}",
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: - is failed -> has failed

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 7]
```diff
@@ -671,13 +674,41 @@ public void run() {
...
+            }
         }
     }
 
     private static void safeFail(ChannelPromise promise, Throwable cause) {
         if (!(promise instanceof VoidChannelPromise) && !promise.tryFailure(cause)) {
-            logger.warn("Failed to mark a promise as failure because it's done already: {}", promise, cause);
+            Throwable err = promise.cause();
+            if (err == null) {
+                logger.warn("Failed to mark a promise as failure because it's succeed already: {}", promise, cause);
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: - 's succeed -> has succeeded

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 8]
```diff
@@ -671,13 +674,41 @@ public void run() {
...
+            } else {
+                logger.warn("Failed to mark a promise as failure because it's failed already: {} with cause {}",
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: - 's failed -> has failed

Comment (trustin) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: - `with cause {}` -> `, unnotified cause: {}`

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 6 above)
Comment (trustin) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: - `with cause {}` -> `, unnotified cause: {}`

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (61ae4a3a7815ea900f4965c2133e7b568c1689c8) and 4.1 (d6022772047bf0e5b4031bcaab5608d1f1342cc0)

**Pull request closed without merging**